### PR TITLE
[enhance] improve dict in-predicate evaluate

### PR DESCRIPTION
**Pull request by englefly**
# Proposed changes
when column is dict encoded, in_predicate::evaluate() put dict code of IN_LIST into a set, and then compare the selected codes with column cells. The selected dict codes are put into a map, by which the evaluate time is O(logN)
It is better to use a std::vector<bool> to indicated if a dict word is in IN_LIST, by which the evaluate time shrinks to O(1).

test evn: ssb_flat 5G 
test sql: simplified from ssb q.3.3
SELECT count(c_city) FROM lineorder_flat WHERE C_CITY in ('UNITED KI1','UNITED KI5');

ShortPredEvalTime is decreased from 240ms to 170ms by average.

在测试环境中 ssb100-flat， 3BE q3.3 从291ms下降到248ms

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/in_list_predicate.cpp [hunk 1]
```diff
@@ -161,12 +162,14 @@ IN_LIST_PRED_COLUMN_BLOCK_EVALUATE(NotInListPredicate, ==)
                         reinterpret_cast<vectorized::ColumnDictionary<vectorized::Int32>&>(      \
                                 column);                                                         \
                 auto& data_array = dict_col.get_data();                                          \
-                auto dict_codes = dict_col.find_codes(_values);                                  \
+                std::vector<bool> selected;                                                  \
+                size_t dict_word_num = 0;                                                    \
```
Comment (yiguolei) [inline be/src/olap/in_list_predicate.cpp:166]: dict_word_num is never used?

Comment (englefly) [reply be/src/olap/in_list_predicate.cpp:166]: I am not sure if there should be an assert(cell_value<dict_word_num).
I will add this assert.

File: be/src/olap/in_list_predicate.cpp [hunk 2]
```diff
@@ -132,13 +132,14 @@ IN_LIST_PRED_COLUMN_BLOCK_EVALUATE(NotInListPredicate, ==)
...
+                    std::vector<bool> selected;                                                  \
+                    size_t dict_word_num = 0;                                                    \
+                    nested_col_ptr->find_codes(_values, selected, dict_word_num);                \
                     for (uint16_t i = 0; i < *size; i++) {                                       \
                         uint16_t idx = sel[i];                                                   \
                         sel[new_size] = idx;                                                     \
                         const auto& cell_value = data_array[idx];                                \
-                        bool ret = !null_bitmap[idx] &&                                          \
-                                   (dict_codes.find(cell_value) OP dict_codes.end());            \
+                        bool ret = !null_bitmap[idx] && (selected[cell_value] OP false);                    \
```
Comment (HappenLee) [inline be/src/olap/in_list_predicate.cpp:142]: code format of `\`

File: be/src/olap/in_list_predicate.cpp (hunk 1 above)
Comment (englefly) [reply be/src/olap/in_list_predicate.cpp]: 已经去掉了 dict_word_num 参数

**Review by yiguolei — DISMISSED** [review]
LGTM

File: be/src/vec/columns/column_dictionary.h [hunk 3]
```diff
@@ -258,9 +258,9 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
 
     uint32_t get_hash_value(uint32_t idx) const { return _dict.get_hash_value(_codes[idx]); }
 
-    phmap::flat_hash_set<int32_t> find_codes(
-            const phmap::flat_hash_set<StringValue>& values) const {
-        return _dict.find_codes(values);
+    void find_codes(const phmap::flat_hash_set<StringValue>& values,
+                    std::vector<bool>& selected) const {
+        return _dict.find_codes(values, selected);
```
Comment (zenoyang) [inline be/src/vec/columns/column_dictionary.h:263]: `find_codes` passes the result `selected` by reference, and not by `return`. This is not consistent with other interfaces, such as `find_code`, `find_code_by_bound`. I think it's better to keep it uniform.

File: be/src/olap/in_list_predicate.cpp [hunk 4]
```diff
@@ -161,12 +162,14 @@ IN_LIST_PRED_COLUMN_BLOCK_EVALUATE(NotInListPredicate, ==)
...
                 auto& data_array = dict_col.get_data();                                          \
-                auto dict_codes = dict_col.find_codes(_values);                                  \
+                std::vector<bool> selected;                                                      \
+                dict_col.find_codes(_values, selected);                                          \
                 for (uint16_t i = 0; i < *size; i++) {                                           \
                     uint16_t idx = sel[i];                                                       \
                     sel[new_size] = idx;                                                         \
                     const auto& cell_value = data_array[idx];                                    \
-                    auto result = (dict_codes.find(cell_value) OP dict_codes.end());             \
+                    assert(cell_value < selected.size());                                        \
```
[the remark this answers is not in the document]
Comment (yiguolei) [reply be/src/olap/in_list_predicate.cpp:171]: Yes, doris uses dcheck instead of assert.

[the remark this answers is not in the document]
Comment (englefly) [reply be/src/olap/in_list_predicate.cpp:171]: assert only works in debug mode。 It will not affect the performance of release version

**Pull request closed without merging**